File: Valid_Parentheses.cpp
/*20. Valid Parentheses  QuestionEditorial Solution  My Submissions
Total Accepted: 136605 Total Submissions: 439636 Difficulty: Easy
Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.

Hide Company Tags Google Airbnb Facebook Twitter Zenefits Amazon Microsoft Bloomberg
Hide Tags Stack String
Hide Similar Problems (M) Generate Parentheses (H) Longest Valid Parentheses (H) Remove Invalid Parentheses
*/




/*use stack to store parentheses
if it is  ( [ {  push them in
else pop them, and check*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    bool isValid(string s) {
        stack<char> s1;
        
        for(char c:s){
            if(c=='(' || c=='{' || c=='[') s1.push(c);
            else if(c==')' && !s1.empty() && s1.top()=='(') s1.pop();
            else if(c=='}' && !s1.empty() && s1.top()=='{') s1.pop();
            else if(c==']' && !s1.empty() && s1.top()=='[') s1.pop();
            else  return false;
        }
        
        return s1.empty();
    }
};

class Solution {
public:
    bool isValid(string s) {
        stack<char> s1;
        for(char &c : s){
            switch(c){
                case '(': //s1.push(c); break;
                case '{': //s1.push(c); break;
                case '[': s1.push(c); break;
                case ')': if(s1.empty() || s1.top()!='(') return false; else s1.pop(); break;
                case '}': if(s1.empty() || s1.top()!='{') return false; else s1.pop(); break;
                case ']': if(s1.empty() || s1.top()!='[') return false; else s1.pop(); break;
                default: ;
            }
        }
        return s1.empty();
    }
};






/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public boolean isValid(String s) {
        Stack<Character> s1 = new Stack<>();
        for(char c:s.toCharArray()){
            if(c=='(' || c=='{' || c=='[' ) s1.push(c);
            else if(c==')' && !s1.empty() && s1.peek()=='(') s1.pop(); //remember !empty
            else if(c=='}' && !s1.empty() && s1.peek()=='{') s1.pop();
            else if(c==']' && !s1.empty() && s1.peek()=='[') s1.pop();
            else return false; //need to return false like []} situation
        }
        
        return s1.empty();
    }
}

File: Valid_Parentheses.java
/*
 Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

 An input string is valid if:

 Open brackets must be closed by the same type of brackets.
 Open brackets must be closed in the correct order.
 Note that an empty string is also considered valid.

 Example 1:

 Input: "()"
 Output: true
 Example 2:

 Input: "()[]{}"
 Output: true
 Example 3:

 Input: "(]"
 Output: false
 Example 4:

 Input: "([)]"
 Output: false
 Example 5:

 Input: "{[]}"
 Output: true

 Hint
 An interesting property about a valid parenthesis expression is that a sub-expression of a valid expression should also be a valid expression. (Not every sub-expression) e.g.
 { { } [ ] [ [ [ ] ] ] } is VALID expression
           [ [ [ ] ] ]    is VALID sub-expression
   { } [ ]                is VALID sub-expression
 Can we exploit this recursive structure somehow?

 What if whenever we encounter a matching pair of parenthesis in the expression, we simply remove it from the expression? This would keep on shortening the expression. e.g.
 { { ( { } ) } }
       |_|
 { { (      ) } }
     |______|
 { {          } }
   |__________|
 {                }
 |________________|
 VALID EXPRESSION!

 The stack data structure can come in handy here in representing this recursive structure of the problem. We can't really process this from the inside out because we don't have an idea about the overall structure. But, the stack can help us process this recursively i.e. from outside to inwards.
*/

/*
Solution 2:
stack
optimize solution 1

Time Complexity: O(n)
Space Complexity: O(n) - all char are left or right
*/
class Solution {
    public boolean isValid(String s) {
        if (s.length() == 0) {
            return true;
        }

        Map<Character, Character> dic = new HashMap<>();
        dic.put('}', '{');
        dic.put(']', '[');
        dic.put(')', '(');

        Stack<Character> leftStack = new Stack<>();
        for(int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (dic.containsKey(c)) {
                // c is right bracket
                char leftKey = dic.get(c);
                if (leftStack.size() > 0 && leftKey == leftStack.peek()) {
                    leftStack.pop();
                } else {
                    return false;
                }
            } else {
                // c is left bracket
                leftStack.push(c);
            }
        }
        return leftStack.size() == 0;
    }
}